framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -240,48 +246,50 @@ private AnnotationMirror createBooleanAnnotationMirror(List<Boolean> values) {
             Node rightNode,
             NumbericalBinaryOps op,
             TransferInput<CFValue, CFStore> p) {
-        List<? extends Number> lefts = getNumericalValues(leftNode, p);
-        List<? extends Number> rights = getNumericalValues(rightNode, p);
         List<Number> resultValues = new ArrayList<>();
-        for (Number left : lefts) {
-            NumberMath<?> nmLeft = NumberMath.getNumberMath(left);
-            for (Number right : rights) {
-                switch (op) {
-                    case ADDTION:
-                        resultValues.add(nmLeft.plus(right));
-                        break;
-                    case DIVISION:
-                        resultValues.add(nmLeft.divide(right));
-                        break;
-                    case MULPLICATION:
-                        resultValues.add(nmLeft.times(right));
-                        break;
-                    case REMAINDER:
-                        resultValues.add(nmLeft.remainder(right));
-                        break;
-                    case SUBTRACTION:
-                        resultValues.add(nmLeft.minus(right));
-                        break;
-                    case SHIFT_LEFT:
-                        resultValues.add(nmLeft.shiftLeft(right));
-                        break;
-                    case SIGNED_SHIFT_RIGHT:
-                        resultValues.add(nmLeft.signedSiftRight(right));
-                        break;
-                    case UNSIGNED_SHIFT_RIGHT:
-                        resultValues.add(nmLeft.unsignedSiftRight(right));
-                        break;
-                    case BITWISE_AND:
-                        resultValues.add(nmLeft.bitwiseAnd(right));
-                        break;
-                    case BITWISE_OR:
-                        resultValues.add(nmLeft.bitwiseOr(right));
-                        break;
-                    case BITWISE_XOR:
-                        resultValues.add(nmLeft.bitwiseXor(right));
-                        break;
-                    default:
-                        throw new UnsupportedOperationException();
+        if (!isIntRange(leftNode, p) && !isIntRange(rightNode, p)) {
+            List<? extends Number> lefts = getNumericalValues(leftNode, p);
+            List<? extends Number> rights = getNumericalValues(rightNode, p);
+            for (Number left : lefts) {
+                NumberMath<?> nmLeft = NumberMath.getNumberMath(left);
+                for (Number right : rights) {
+                    switch (op) {
+                        case ADDTION:
+                            resultValues.add(nmLeft.plus(right));
```

Comment:
What do you think of modifying `NuericalBinaryOps` and adding proper methods there.
Then, instead of the ugly large switch statement, you could have a single line:
`resultValues.add( op.handleBinary(nmLeft, right) )`.
That would be a bit larger change, but the result would be much more object oriented.
Oh, but that would also mean changing what `NumberMath` does. Hmm, let's not do this then...